src/mpid/ch4/shm/posix/eager/iqueue/iqueue_init.h

```diff
@@ -0,0 +1,136 @@
+/* -*- Mode: C; c-basic-offset:4 ; indent-tabs-mode:nil ; -*- */
+/*
+ *  (C) 2006 by Argonne National Laboratory.
+ *      See COPYRIGHT in top-level directory.
+ *
+ *  Portions of this code were written by Intel Corporation.
+ *  Copyright (C) 2011-2017 Intel Corporation.  Intel provides this material
+ *  to Argonne National Laboratory subject to Software Grant and Corporate
+ *  Contributor License Agreement dated February 8, 2012.
+ */
+#ifndef POSIX_EAGER_IQUEUE_INIT_H_INCLUDED
+#define POSIX_EAGER_IQUEUE_INIT_H_INCLUDED
+
+#include "iqueue_types.h"
+
+/*
+=== BEGIN_MPI_T_CVAR_INFO_BLOCK ===
+
+cvars:
+    - name        : MPIR_CVAR_CH4_SHM_POSIX_IQUEUE_NUM_CELLS
+      category    : CH4
+      type        : int
+      default     : 32
+      class       : device
+      verbosity   : MPI_T_VERBOSITY_USER_BASIC
+      scope       : MPI_T_SCOPE_ALL_EQ
+      description : >-
+        The number of cells used for the depth of the iqueue.
+
+=== END_MPI_T_CVAR_INFO_BLOCK ===
+*/
+
+MPL_STATIC_INLINE_PREFIX int MPIDI_POSIX_eager_init(int rank, int size)
+{
+    int mpi_errno = MPI_SUCCESS;
```

Comment:
@raffenet @yfguo @hzhou fbox_init.h and iqueue_init.h are still header files. Do we plan to replace them with .c?

Reply:
I started a PR to uninline fastbox but it proved more complicated than expected https://github.com/pmodels/mpich/pull/4209

Reply:
I prefer making it a .c file rather than .h.

Reply:
Agree. Let's change it in this PR. @gcongiu It will be good if we first fix the issue in fbox. I am checking that PR now.

Reply:
I still have reservation on the contention concerns when multiple processes sending to a single process -- a very common application pattern. We need decide whether to abandon fastbox or workout a dynamic scheme that use both fastbox and iqueue at runtime before move forward with this PR.

Reply:
Then senders and receiver also contend on the queue, right?

Reply:
@hzhou Yes, a `hybrid` (or dynamic) version that combines both fbox and iqueue would be ideal (that is what ch3 does). But we do not have time to work on it. And in fact the difference between the `hybrid` version and the `iqueue-only` version is very small (i.e., iqueue has worse small message latency). @pavanbalaji actually pointed out that there are some applications care about **small message latency**, but I am not sure how much difference we could see in application level. 

In summary, replacing `fbox` with `iqueue` will definitely give us benefit, and @gcongiu can go ahead with his GPU work which relies on `iqueue` (this is why we are pushing this PR now). We need more time to investigate the benefit of the `hybrid` version and I'd suggest we put its development in future plan. We can create a ticket to track this issue.

Reply:
How does GPU work depends on iqueue?

Reply:
An example is that when IPC is not available, the sender has to copy the data from GPU to a temporary host buffer. A cell buffer in iqueue is a good option. The data transfer of an intranode send-receive can be implemented as:
gpu_buf on P0 -> cell buffer -> gpu_buf on P1

If we do the above thing with `fbox`, the performance will be likely always worse than that with `iqueue`. E.g., when chunking a large message into multiple pieces in order to fit into the fbox buffer.

As we have a milestone for GPU work, we need to prioritize the tasks. The most effective & less time option is to use iqueue as the default, and add the GPU functionality on top of it. @gcongiu please feel free to comment if I miss anything.

Reply:
Correct, we always planned to use shared memory for intranode fallback when GPUDirect P2P is not usable thus, as Min said, the easiest way of doing this is through iqueues. As @pavanbalaji also noted this might be even faster than using IPC as memory handles have to be created to exchange addresses among processes before actual copy of data can start. Since creating these handle is very expensive and we don't know the location of the user buffer beforehand, iqueues might provide an effective intranode communication for the general case, not only fallback.

Reply:
> If we do the above thing with `fbox`, the performance will be likely always worse than that with `iqueue`. E.g., when chunking a large message into multiple pieces in order to fit into the fbox buffer.

You are chunking a large message into multiple pieces with cells too, right? The argument of performance is the same argument for general large messages, right? I am not saying it is wrong, but it seems to be an orthogonal argument.

There are benefits If we can keep the lower-layer transport orthogonal to the upper layer. Certainly one transport may perform better than another, but that is different from making GPU depend on one particular transport, as that will constrain the code in both layers.

Reply:
The GPU intranode fallback does not depend on any transport. It just depends on shared memory. The choice of iqueue is just for better performance. This is my understanding.